src/mscorlib/src/System/Lazy.cs

```diff
@@ -322,147 +516,12 @@ public T Value
         {
             get
             {
-                Boxed boxed = null;
-                if (m_boxed != null )
-                {
-                    // Do a quick check up front for the fast path.
-                    boxed = m_boxed as Boxed;
-                    if (boxed != null)
-                    {
-                        return boxed.m_value;
-                    }
-
-                    LazyInternalExceptionHolder exc = m_boxed as LazyInternalExceptionHolder;
-                    Debug.Assert(exc != null);
-                    exc.m_edi.Throw();
-                }
-
-                // Fall through to the slow path.
-                return LazyInitValue();
+                var state = _state;
+                if (state == null)
+                    return _value;
+                return LazyGetValue(state);
```

Comment:
I implemented this, but it isn't right I don't think.

Hmmm. There is a race condition here (well in the previous as well I think) with the interaction that this on in PublicationOnly and ExecutionAndPublication who the state variable. Anyway I need to think through the implications. but don't have time at the moment. Maybe tonight.